File: Data Structures & Algorithms/subsets/submission-2.py
from typing import List

class Solution:
    def subsets(self, nums: List[int]) -> List[List[int]]:
        def dfs(nums: List[int], index: int, curr: List[int]):
            temp = curr.copy()
            if index == len(nums):
                return [temp]
            temp.append(nums[index])
            ls1 = dfs(nums, index + 1, temp)
            temp.pop()
            ls2 = dfs(nums, index + 1, temp)
            ls1.extend(ls2)
            return ls1
        
        return dfs(nums, 0, [])

File: Data Structures & Algorithms/binary-search/submission-2.cpp
class Solution {
public:
    int search(vector<int>& nums, int target) {
        int n = nums.size();
        int bot = 0;
        int top = n - 1;
        while (bot != top) {
            int index = (bot + top) / 2;
            if (nums[index] == target) {
                return index;
            }
            else if (nums[index] < target) {
                bot = index + 1;
            }
            else {
                top = index;
            }
        }
        if (target == nums[bot]) {
            return bot;
        }
        return -1;
    }
};
